Sources/Moya/Task.swift

```diff
@@ -3,39 +3,84 @@ import Foundation
+        public let encodable: Encodable
+        /// The encoder to be used to encode `encodable` into an `URLRequest`.
+        public let encoder: ParameterEncoder
 
-    /// A requests body set with encoded parameters.
-    case requestParameters(parameters: [String: Any], encoding: ParameterEncoding)
+        internal init(encodable: Encodable, encoder: ParameterEncoder) {
+            self.encodable = encodable
+            self.encoder = encoder
+        }
+    }
 
-    /// A requests body set with data, combined with url parameters.
-    case requestCompositeData(bodyData: Data, urlParameters: [String: Any])
+    /// A container for some parameters to be encoded into an `URLRequest`'s url's query string using a specific encoder.
+    public final class URLParams: Parameters {
 
-    /// A requests body set with encoded parameters combined with url parameters.
-    case requestCompositeParameters(bodyParameters: [String: Any], bodyEncoding: ParameterEncoding, urlParameters: [String: Any])
+        /// The designated method to initialize the `URLParams`.
+        /// - Parameters:
+        ///   - encodable: The parameters to be encoded into the url's query string.
+        ///   - encoder: The encoder to be used to encode the encodable into the url's query string.
+        public required init(_ encodable: Encodable, encoder: URLEncodedFormEncoder = URLEncodedFormEncoder()) {
+            super.init(encodable: encodable, encoder: URLEncodedFormParameterEncoder(encoder: encoder, destination: .queryString))
+        }
+    }
 
-    /// A file upload task.
-    case uploadFile(URL)
+    /// A container for some parameters to be encoded into an `URLRequest` using a specific encoder.
+    public final class CustomParams: Parameters {
 
-    /// A "multipart/form-data" upload task.
-    case uploadMultipart([MultipartFormData])
+        /// The designated method to initialize `CustomParams`.
+        /// - Parameters:
+        ///   - encodable: The parameters to be encoded into the `URLRequest`.
+        ///   - encoder: The encoder to be used to encode `encodable` into an `URLRequest`.
+        ///   It **must not** be a `URLEncodedFormParameterEncoder` (use `BodyParams.urlEncoded`
+        ///   or `URLParams` instead) or `JSONParameterEncoder`(use `BodyParams.json` instead).
+        ///   If this is the case, a `MoyaError.encodableMapping` will be raised.
+        public required init(_ encodable: Encodable, encoder: ParameterEncoder) throws {
+            // CustomParams should only be used when `URLEncodedFormParameterEncoder` and `JSONParameterEncoder`
+            // are not enough. To enforce usage of BodyParams or URLParams, let's check the type of the given encoder
+            // to make sure CustomParams is correctly used.
+            if encoder is JSONParameterEncoder
+            || encoder is URLEncodedFormParameterEncoder {
+                throw MoyaError.taskParametersUsage(type(of: encoder))
+            }
```

Comment:
do we need to do this check? you could create your own encoding with different class name and still pass something like `JSONParameterEncoder` anyways.

Reply:
There is indeed the possibility to workaround the check is you really want to use params in a way not preferred by Moya. If so then the user explicitely says "I know what i'm doing" and should face possible strange behaviors.

If we remove this check, users would have the possibility to express their parameters either with `BodyParams` or `URLParams` and `CustomParams`, which I think is not a good API. We then would have to unify the 3 different params types into only one.
Please see https://github.com/Moya/Moya/pull/1936#issuecomment-569953087 i've studied impacts of such an update.

Reply:
ah very good explanation - thanks! it is indeed a hard one, was thinking about it for a while and not many choices left here... 

one thing that I had in mind was - what if we had a classic `Parameters` struct that would be suitable for `request`/`download` type and `UploadParameters` struct only for `upload` type? do you think it would eliminate most of the problems and not introduce new ones?

Reply:
I don't think we have problems with the actual code. I chose to help the user by providing a type safe API, at the lost of more code complexity. We can have a less complex code but the API would be less appealing to the user, and may allow some unexpected behaviors. I explored this second option in my comment above.

If your 2 structs idea simplifies the code while still being as appealing as the current solution I would gladly explore it. But right know I struggle to understand what properties would those 2 structs contain. Can you provide an example ?

Reply:
Yeah I didn't think about the actual code as problems, rather the check for types is something I would love to avoid if possible so that's why I'm thinking out loud. About the `Parameters`/`UploadParamaters` - I was thinking that the `Parameters` struct could allow for all properties (body/url/whatever) and `UploadParameters` would allow only for url ones?

Reply:
I don't see how such a distinction would suppress the current type check in CustomParams.